### PR TITLE
Rebase sycl-nliber2 into develop

### PR DESCRIPTION
**Pull request by nliber**
Work in progress

### CODE REVIEW

File: Makefile.kokkos [hunk 1]
```diff
@@ -5,7 +5,7 @@ KOKKOS_VERSION_MINOR = 1
 KOKKOS_VERSION_PATCH = 1
 KOKKOS_VERSION = $(shell echo $(KOKKOS_VERSION_MAJOR)*10000+$(KOKKOS_VERSION_MINOR)*100+$(KOKKOS_VERSION_PATCH) | bc)
 
-# Options: Cuda,HIP,ROCm,OpenMP,Pthread,Serial
+# Options: Cuda,HIP,SYCLROCm,OpenMP,Pthread,Serial
```
Comment (dalg24) [inline Makefile.kokkos]:
```suggestion
# Options: Cuda,HIP,SYCL,OpenMP,Pthread,Serial
```

File: Makefile.kokkos [hunk 2]
```diff
@@ -340,9 +361,9 @@ KOKKOS_INTERNAL_USE_ARCH_NVIDIA := $(shell expr $(KOKKOS_INTERNAL_USE_ARCH_KEPLE
                                               + $(KOKKOS_INTERNAL_USE_ARCH_MAXWELL53) \
                                               + $(KOKKOS_INTERNAL_USE_ARCH_PASCAL61)  \
                                               + $(KOKKOS_INTERNAL_USE_ARCH_PASCAL60)  \
-                                              + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA70)   \
-                                              + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA72)   \
-                                              + $(KOKKOS_INTERNAL_USE_ARCH_TURING75)  \
+											  + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA70) \
+											  + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA72) \
+											  + $(KOKKOS_INTERNAL_USE_ARCH_TURING75) \
```
Comment (dalg24) [inline Makefile.kokkos]: Undo that change
```suggestion
                                               + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA70) \
                                               + $(KOKKOS_INTERNAL_USE_ARCH_VOLTA72) \
                                               + $(KOKKOS_INTERNAL_USE_ARCH_TURING75) \
```

File: Makefile.kokkos [hunk 3]
```diff
@@ -598,15 +623,15 @@ endif
...
-      KOKKOS_TPL_LIBRARY_DIRS += $(HWLOC_PATH)/lib
-    endif
-    KOKKOS_LIBS += -lhwloc
-    KOKKOS_TPL_LIBRARY_NAMES += hwloc
+    KOKKOS_CXXLDFLAGS += -L$(HWLOC_PATH)/lib
+    KOKKOS_TPL_INCLUDE_DIRS += $(HWLOC_PATH)/include
+    KOKKOS_TPL_LIBRARY_DIRS += $(HWLOC_PATH)/lib
+  endif
+  KOKKOS_LIBS += -lhwloc
+  KOKKOS_TPL_LIBRARY_NAMES += hwloc
```
Comment (dalg24) [inline Makefile.kokkos]: Undo this too

File: Makefile.kokkos [hunk 4]
```diff
@@ -619,15 +644,15 @@ endif
...
-      KOKKOS_TPL_LIBRARY_DIRS += $(MEMKIND_PATH)/lib
-    endif
-    KOKKOS_LIBS += -lmemkind -lnuma
-    KOKKOS_TPL_LIBRARY_NAMES += memkind numa
+    KOKKOS_CXXLDFLAGS += -L$(MEMKIND_PATH)/lib
+    KOKKOS_TPL_INCLUDE_DIRS += $(MEMKIND_PATH)/include
+    KOKKOS_TPL_LIBRARY_DIRS += $(MEMKIND_PATH)/lib
+  endif
+  KOKKOS_LIBS += -lmemkind -lnuma
+  KOKKOS_TPL_LIBRARY_NAMES += memkind numa
```
Comment (dalg24) [inline Makefile.kokkos]: Undo this

File: core/src/Kokkos_SYCL.hpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+
+  uint32_t impl_instance_id() const noexcept { return 0; }
+
+  //@}
+  //------------------------------------
+  //! \name Functions that all Kokkos devices must implement.
+  //@{
+
+  KOKKOS_INLINE_FUNCTION static int in_parallel() {
+#if defined(__HCC_ACCELERATOR__)
```
Comment (dalg24) [inline core/src/Kokkos_SYCL.hpp:112]: Where is that coming from?

File: core/src/Kokkos_SYCL.hpp [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  static int concurrency();
+  static const char* name();
+
+  inline Impl::SYCLInternal* impl_internal_space_instance() const {
+    return m_space_instance;
+  }
+
+ private:
+  Impl::SYCLInternal* m_space_instance;
+};  // namespace Experimental
```
Comment (dalg24) [inline core/src/Kokkos_SYCL.hpp:182]:
```suggestion
};
```

File: core/src/Kokkos_View.hpp [hunk 7]
```diff
@@ -1422,26 +1422,17 @@ class View : public ViewTraits<DataType, Properties...> {
...
+  View& operator=(const View& rhs) = default;
 
-  KOKKOS_INLINE_FUNCTION
-  View& operator=(View&& rhs) {
-    m_track = std::move(rhs.m_track);
-    m_map   = std::move(rhs.m_map);
-    return *this;
-  }
+  KOKKOS_DEFAULTED_FUNCTION
+  View& operator=(View&& rhs) = default;
```
Comment (dalg24) [inline core/src/Kokkos_View.hpp:1435]: Please open a separate pull request with these changes to View.

File: core/src/SYCL/Kokkos_SYCL_Error.hpp [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+inline void sycl_internal_safe_call( syclError_t e , const char * name, const char * file = NULL, const int line = 0)
+{
+  if ( syclSuccess != e ) { sycl_internal_error_throw( e , name, file, line ); }
+}
+*/
+}
+}
+#define SYCL_SAFE_CALL( call )  \
+        call
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Error.hpp]: Did you want to keep this?  This is used in backends where the runtime use error codes.  My understanding is SYCL uses exceptions.

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+#include <SYCL/Kokkos_SYCL_Error.hpp>
+#include <CL/sycl.hpp>
+
+template<class T>
+class kokkos_sycl_functor;
+
+#define CONCAT_IMPL( x, y ) x##y
+#define MACRO_CONCAT( x, y ) CONCAT_IMPL( x, y )
+#define CGH_PARALLEL_FOR cgh.parallel_for<class MACRO_CONCAT( kokkos_sycl_functor, __COUNTER__ )>
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: These probably need to be prefixed with `KOKKOS_IMPL`

File: core/src/impl/Kokkos_Error.hpp [hunk 10]
```diff
@@ -55,6 +55,12 @@
 #include <HIP/Kokkos_HIP_Abort.hpp>
 #endif
 
+// SYCL defines a macro abort() when using nvcc, thus we get a collision with the abort function
+// defined here
+#ifdef KOKKOS_ENABLE_SYCL
+  #undef abort
+#endif
```
Comment (dalg24) [inline core/src/impl/Kokkos_Error.hpp:62]: Yuk

File: generate_makefile.bash [hunk 11]
```diff
@@ -355,6 +359,14 @@ if [[ ${KOKKOS_DEVICE_CMD} == *Kokkos_ENABLE_HPX* ]]; then
    fi
 fi
 
+## if SYCL is enabled, we need to enforce cxx standard = 17
+if [[ ${KOKKOS_DEVICE_CMD} == *Kokkos_ENABLE_SYCL* ]]; then
+   if [ "${KOKKOS_CXX_STANDARD}" == "" ] || [ ${#KOKKOS_CXX_STANDARD} -lt 17 ]; then
+      echo CXX Standard must be 17 or higher for SYCL to work.
+      KOKKOS_CXX_STANDARD=17
+   fi
+fi
```
Comment (dalg24) [inline generate_makefile.bash:368]: We would need to require C++17 in CMake as well.
@nliber can you remind me why we can't use C++14?

File: core/src/SYCL/Kokkos_SYCL_Space.cpp [hunk 12]
```diff
@@ -0,0 +1,1028 @@
...
+void USM_memcpy(void* dst, const void* src, size_t n) {
+  USM_memcpy(Kokkos::Experimental::Impl::SYCLInternal::singleton(), dst, src,
+             n);
+}
+}  // namespace
+
+DeepCopy<Kokkos::Experimental::SYCLSpace, Kokkos::Experimental::SYCLSpace,
+         Kokkos::Experimental::SYCL>::DeepCopy(void* dst, const void* src,
+                                               size_t n) {
+  //  syclMemcpy(dst , src , n , syclMemcpyDefault );
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Space.cpp:90]: I would want us to fix or raise an error.

File: core/src/SYCL/Kokkos_SYCL_Space.cpp [hunk 13]
```diff
@@ -0,0 +1,1028 @@
...
+namespace Kokkos {
+namespace Experimental {
+
+SYCLSpace::SYCLSpace() : m_device(SYCL().sycl_device()) {}
+
+SYCLHostPinnedSpace::SYCLHostPinnedSpace() {}
+
+void* SYCLSpace::allocate(const size_t arg_alloc_size) const {
+  void* ptr = NULL;
+  // syclMalloc( &ptr, arg_alloc_size );
```
Comment (dalg24) [inline core/src/SYCL/Kokkos_SYCL_Space.cpp:287]: Why did you comment?

File: core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp (hunk 9 above)
Comment (nliber) [reply core/src/SYCL/Kokkos_SYCL_KernelLaunch.hpp]: Will do.

File: core/src/Kokkos_SYCL.hpp (hunk 5 above)
Comment (nliber) [reply core/src/Kokkos_SYCL.hpp:112]: Bad merge?  Anywaywill fix.

File: generate_makefile.bash (hunk 11 above)
Comment (masterleinad) [reply generate_makefile.bash:368]: So far I have not seen that I need to explicitly have to provide a standard flag.

Comment (nliber) [reply generate_makefile.bash:368]: SYCL 2020 is requiring C++17.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Most of the current compilation errors look like
```
/var/jenkins/workspace/Kokkos/core/src/Kokkos_SYCL.hpp:70:3: error: static_assert failed due to requirement 'std::is_trivially_copyable<Kokkos::Impl::ParallelFor<Kokkos::Impl::ViewValueFunctor<Kokkos::Experimental::SYCL, double, true>, Kokkos::RangePolicy<Kokkos::Experimental::SYCL, Kokkos::IndexType<long>>, Kokkos::Experimental::SYCL>>::value'

  static_assert(std::is_trivially_copyable<Obj>::value);
```
we need to decide how to handle that.

**Comment by masterleinad** [comment]
I am trying to just disable as many tests as necessary for this to compile such that we can move forward here.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
OK. This is finally passing.

**Pull request closed without merging**